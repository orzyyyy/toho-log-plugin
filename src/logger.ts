import chalk from "chalk";

export interface WorkProps {
  mean: string;
  word: string;
}

const log = (text: string, callbackData?: WorkProps[]) => {
  if (callbackData && callbackData.length != 0) {
    const random = ~~(Math.random() * callbackData.length);
    const item = callbackData[random];
    console.log(
      chalk.greenBright(`${text}   =>   ${item.word}  ${item.mean} `)
    );
  } else {
    console.log(chalk.greenBright(text));
  }
};
const error = (text: string) => console.log(chalk.red(text));
const warn = (text: string) => console.log(chalk.yellowBright(text));
const info = (text: string) => console.log(chalk.cyanBright(text));

let successCount = 0;

const logInfo = (
  err: any,
  stats: any,
  isDev: boolean,
  callbackData?: WorkProps[]
) => {
  if (err) {
    error(err.stack || err);
    if (err.details) {
      error(err.details);
    }
    return;
  }
  const statsInfo = stats.toJson();
  if (stats.hasErrors()) {
    for (let item of statsInfo.errors) {
      error(item);
    }
    error("\n  少女以为能神穿，奈何缠在腰间的香火钱太多，戳之，卒 (°□°；) \n");
    return;
  }

  if (stats.hasWarnings()) {
    let i = 0;
    for (let item of statsInfo.warnings) {
      i++;
      warn(i + ". " + item.message + " \n");
    }
    if (!isDev) {
      warn("\n  虽然有些烦恼，但少女还是去和风车战斗了\n");
      logForProd();
      return;
    }
  }
  if (!isDev) {
    logForProd();
  } else {
    log(
      `♪(^∇^*)♪(^∇^*)♪(^∇^*) 少女第 ${++successCount} 次捡到钱了 ♪(^∇^*)♪(^∇^*)♪(^∇^*)`,
      callbackData
    );
  }
};

const logForProd = () => {
  log("\n  少女抬首，风车空转。影动风移，浮沉氤氲。若一去不回，便一去不回罢\n");
  log(
    "  铁血的热血的冷血的可笑的可悲的可爱的可敬的少女死去了，但好像又活了过来\n"
  );
  info("  然后少女去寻找自己的诗和苟且了 ╮(╯_╰)╭\n");
};

export { log, error, warn, info, logInfo };
