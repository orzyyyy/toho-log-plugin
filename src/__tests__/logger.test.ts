import { log, error, warn, info, logInfo } from "../logger";
import chalk from "chalk";

describe("logger", () => {
  const logSpy = jest.spyOn(console, "log").mockImplementation(() => {});

  afterAll(() => {
    logSpy.mockRestore();
  });

  it("log", () => {
    log("test", [{ mean: "mean", word: "word" }]);
    expect(logSpy).toHaveBeenCalledWith(
      chalk.greenBright("test   =>   word  mean ")
    );
    log("test");
    expect(logSpy).toHaveBeenCalledWith(chalk.greenBright("test"));
  });

  it("error", () => {
    error("test");
    expect(logSpy).toHaveBeenCalledWith(chalk.red("test"));
  });

  it("warn", () => {
    warn("test");
    expect(logSpy).toHaveBeenCalledWith(chalk.yellowBright("test"));
  });

  it("info", () => {
    info("test");
    expect(logSpy).toHaveBeenCalledWith(chalk.cyanBright("test"));
  });

  it("logInfo with stack error", () => {
    const result = logInfo({ stack: "stack", details: "details" }, {}, true);
    expect(logSpy).toHaveBeenCalledWith(chalk.red("stack"));
    expect(logSpy).toHaveBeenCalledWith(chalk.red("details"));
    expect(result).toBe(undefined);
  });

  it("logInfo with stats error", () => {
    const result = logInfo(
      null,
      {
        hasErrors: () => true,
        toJson: () => ({ errors: ["error1", "error2"] }),
      },
      true
    );
    expect(logSpy).toHaveBeenCalledWith(chalk.red("error1"));
    expect(logSpy).toHaveBeenCalledWith(chalk.red("error2"));
    expect(logSpy).toHaveBeenCalledWith(
      chalk.red(
        "\n  少女以为能神穿，奈何缠在腰间的香火钱太多，戳之，卒 (°□°；) \n"
      )
    );
    expect(result).toBe(undefined);
  });

  it("logInfo with stats warning", () => {
    const result = logInfo(
      null,
      {
        hasErrors: () => false,
        hasWarnings: () => true,
        toJson: () => ({
          warnings: [{ message: "warnings1" }, { message: "warnings2" }],
        }),
      },
      false
    );
    // expect(logSpy).toHaveBeenCalledWith(chalk.yellowBright("warnings1"));
    // expect(logSpy).toHaveBeenCalledWith(chalk.yellowBright("warnings2"));
    // expect(logSpy).toHaveBeenCalledWith(
    //   chalk.yellowBright("\n  虽然有些烦恼，但少女还是去和风车战斗了\n")
    // );
    expect(logSpy).toHaveBeenCalledTimes(16);
    expect(result).toBe(undefined);
  });

  it("logInfo in dev", () => {
    logInfo(
      null,
      { hasErrors: () => false, hasWarnings: () => false, toJson: () => ({}) },
      true
    );
    expect(logSpy).toHaveBeenCalledWith(
      chalk.greenBright(
        "♪(^∇^*)♪(^∇^*)♪(^∇^*) 少女第 1 次捡到钱了 ♪(^∇^*)♪(^∇^*)♪(^∇^*)"
      )
    );
  });

  it("logInfo in prod", () => {
    logInfo(
      null,
      { hasErrors: () => false, hasWarnings: () => false, toJson: () => ({}) },
      false
    );
    expect(logSpy).toHaveBeenCalledWith(
      chalk.greenBright(
        "\n  少女抬首，风车空转。影动风移，浮沉氤氲。若一去不回，便一去不回罢\n"
      )
    );
    expect(logSpy).toHaveBeenCalledWith(
      chalk.greenBright(
        "  铁血的热血的冷血的可笑的可悲的可爱的可敬的少女死去了，但好像又活了过来\n"
      )
    );
    expect(logSpy).toHaveBeenCalledWith(
      chalk.cyanBright("  然后少女去寻找自己的诗和苟且了 ╮(╯_╰)╭\n")
    );
  });
});
