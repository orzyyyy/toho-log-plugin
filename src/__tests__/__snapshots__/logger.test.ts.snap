// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logger logInfo with stats warning 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[92mtest   =>   word  mean [39m",
    ],
    Array [
      "[92mtest[39m",
    ],
    Array [
      "[31mtest[39m",
    ],
    Array [
      "[93mtest[39m",
    ],
    Array [
      "[96mtest[39m",
    ],
    Array [
      "[31mstack[39m",
    ],
    Array [
      "[31mdetails[39m",
    ],
    Array [
      "[31merror1[39m",
    ],
    Array [
      "[31merror2[39m",
    ],
    Array [
      "[31m[39m
[31m  少女以为能神穿，奈何缠在腰间的香火钱太多，戳之，卒 (°□°；) [39m
[31m[39m",
    ],
    Array [
      "[93m1. undefined [39m
[93m[39m",
    ],
    Array [
      "[93m2. undefined [39m
[93m[39m",
    ],
    Array [
      "[93m[39m
[93m  虽然有些烦恼，但少女还是去和风车战斗了[39m
[93m[39m",
    ],
    Array [
      "[92m[39m
[92m  少女抬首，风车空转。影动风移，浮沉氤氲。若一去不回，便一去不回罢[39m
[92m[39m",
    ],
    Array [
      "[92m  铁血的热血的冷血的可笑的可悲的可爱的可敬的少女死去了，但好像又活了过来[39m
[92m[39m",
    ],
    Array [
      "[96m  然后少女去寻找自己的诗和苟且了 ╮(╯_╰)╭[39m
[96m[39m",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
